jaxopt/_src/gauss_newton.py

```diff
@@ -0,0 +1,165 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""GaussNewton algorithm in JAX."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+
+from dataclasses import dataclass
+
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src import implicit_diff as idf
+from jaxopt._src import loop
+from jaxopt._src import linear_solve
+
+
+class GaussNewtonState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  residual: Any
+  delta: Any
+  error: float
+  gradient: Any
+  jtj: Optional[Any] = None
+  aux: Optional[Any] = None
+
+@dataclass(eq=False)
+class GaussNewton(base.IterativeSolver):
+  """Gauss-Newton nonlinear least-squares solver.
+
+  This solver finds the optimal parameters via a least-squares optimization
+
+    Given the residual function f(x): R^m -> R^n, `gauss_newton` finds a
+    local minimum of the cost function F(x):
+
+    ```
+    argmin_x F(x) = 0.5 * sum(f_i(x)**2), i = 0, ..., m - 1
+    f(x) = func(x, *args, **kwargs)
+    ```
+
+  Attributes:
+    residual_fun: a smooth function of the form ``fun(x, *args, **kwargs)``.
+    maxiter: maximum number of iterations.
+    tol: tolerance.
+    check_bracket: whether to check correctness of the bracketing interval.
+      If True, the method ``run`` cannot be jitted.
+    implicit_diff_solve: the linear system solver to use.
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
+  """
+  residual_fun: Callable
+  maxiter: int = 30
+  tol: float = 1e-5
+  check_bracket: bool = True
+  verbose: bool = False
+  implicit_diff_solve: Optional[Callable] = None
+  has_aux: bool = False
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+  def init_state(self,
+                 init_params: Optional[Any] = None,
+                 *args,
+                 **kwargs) -> GaussNewtonState:
+    """Initialize the solver state.
+
+    Args:
+      init_params: pytree containing the initial parameters.
+      *args: additional positional arguments to be passed to ``fun``.
+      **kwargs: additional keyword arguments to be passed to ``fun``.
+    Returns:
+      state
+    """
+
+    return GaussNewtonState(iter_num=jnp.asarray(0),
+                            error=jnp.asarray(jnp.inf),
+                            residual=jnp.asarray(jnp.inf),
+                            delta=jnp.asarray(jnp.inf),
+                            gradient=jnp.asarray(jnp.inf),
```

Comment:
Note that attributes `residual`, `delta`, `gradient` and `jtj` are scalars when returned by `init_state` but are arrays or pytrees when returned by `update`. This will trigger a jit recompilation because the state will change signature. So if you really want to keep these attributes, it may be better to compute their actual values at `init_params` and return that instead.

Reply:
I implemented your suggestion, very good point, thanks.